library/std/src/thread/scoped.rs

```diff
@@ -0,0 +1,301 @@
+use super::{current, park, Builder, JoinInner, Result, Thread};
+use crate::fmt;
+use crate::io;
+use crate::marker::PhantomData;
+use crate::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};
+use crate::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use crate::sync::Arc;
+
+/// A scope to spawn scoped threads in.
+///
+/// See [`scope`] for details.
+pub struct Scope<'env> {
+    data: ScopeData,
+    env: PhantomData<&'env mut &'env ()>,
+}
+
+/// An owned permission to join on a scoped thread (block on its termination).
+///
+/// See [`Scope::spawn`] for details.
+pub struct ScopedJoinHandle<'scope, T>(JoinInner<'scope, T>);
+
+pub(super) struct ScopeData {
+    n_running_threads: AtomicUsize,
+    a_thread_panicked: AtomicBool,
+    main_thread: Thread,
+}
+
+impl ScopeData {
+    pub(super) fn increment_n_running_threads(&self) {
+        // We check for 'overflow' with usize::MAX / 2, to make sure there's no
+        // chance it overflows to 0, which would result in unsoundness.
+        if self.n_running_threads.fetch_add(1, Ordering::Relaxed) > usize::MAX / 2 {
```

Comment:
The proper way of doing this would be to use a CAS loop (using `fetch_update`) and only updating `n_running_threads` if we are sure there are no overflow. 

Nonetheless, the current method also works reasonably well since you'd need ~2 billion threads suspended at a very specific point to cause UB.

Reply:
Yeah, I'm aware. I figured a CAS loop would be overkill and less readable, which is not worth it for a case that probably never happens anyway. ~2 billion threads having executed fetch_add and not having executed fetch_sub yet means you need one active thread per two bytes in the address space. Storing even just a single return address on the stack for each thread takes more space than that.